core/drivers/stm32_gpio.c

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2019, STMicroelectronics
+ *
+ * STM32 GPIO driver is used as pin controller for stm32mp SoCs.
+ * The driver API is defined in header file stm32_gpio.h.
+ */
+
+#include <assert.h>
+#include <drivers/stm32_gpio.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/generic_boot.h>
+#include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
+#include <stm32_util.h>
+#include <trace.h>
+#include <util.h>
+
+#ifdef CFG_DT
+#include <libfdt.h>
+#endif
+
+#define GPIO_PIN_MAX		15
+
+#define GPIO_MODER_OFFSET	0x00
+#define GPIO_OTYPER_OFFSET	0x04
+#define GPIO_OSPEEDR_OFFSET	0x08
+#define GPIO_PUPDR_OFFSET	0x0c
+#define GPIO_IDR_OFFSET		0x10
+#define GPIO_ODR_OFFSET		0x14
+#define GPIO_BSRR_OFFSET	0x18
+#define GPIO_AFRL_OFFSET	0x20
+#define GPIO_AFRH_OFFSET	0x24
+#define GPIO_SECR_OFFSET	0x30
+
+#define GPIO_ALT_LOWER_LIMIT	0x8
+
+#define GPIO_MODE_MASK		GENMASK_32(1, 0)
+#define GPIO_OSPEED_MASK	GENMASK_32(1, 0)
+#define GPIO_PUPD_PULL_MASK	GENMASK_32(1, 0)
+#define GPIO_ALTERNATE_MASK	GENMASK_32(15, 0)
+
+#define DT_GPIO_BANK_SHIFT	12
+#define DT_GPIO_BANK_MASK	GENMASK_32(16, 12)
+#define DT_GPIO_PIN_SHIFT	8
+#define DT_GPIO_PIN_MASK	GENMASK_32(11, 8)
+#define DT_GPIO_MODE_MASK	GENMASK_32(7, 0)
+
+static unsigned int gpio_lock;
+
+/* Save to output @cfg the current GPIO (@bank/@pin) configuration */
+static void get_gpio_cfg(uint32_t bank, uint32_t pin, struct gpio_cfg *cfg)
+{
+	uintptr_t base = stm32_get_gpio_bank_base(bank);
+	unsigned int clock = stm32_get_gpio_bank_clock(bank);
+
+	stm32_clock_enable(clock);
+
+	cfg->mode = (read32(base + GPIO_MODER_OFFSET) >> (pin << 1)) &
+		     GPIO_MODE_MASK;
+
+	cfg->otype = (read32(base + GPIO_OTYPER_OFFSET) >> pin) & 1;
+
+	cfg->ospeed = (read32(base +  GPIO_OSPEEDR_OFFSET) >> (pin << 1)) &
+		       GPIO_OSPEED_MASK;
+
+	cfg->pupd = (read32(base +  GPIO_PUPDR_OFFSET) >> (pin << 1)) &
+		     GPIO_PUPD_PULL_MASK;
+
+	cfg->od = (read32(base + GPIO_ODR_OFFSET) >> (pin << 1)) & 1;
+
+	if (pin < GPIO_ALT_LOWER_LIMIT)
+		cfg->af = (read32(base + GPIO_AFRL_OFFSET) >> (pin << 2)) &
+			   GPIO_ALTERNATE_MASK;
+	else
+		cfg->af = (read32(base + GPIO_AFRH_OFFSET) >>
+			    ((pin - GPIO_ALT_LOWER_LIMIT) << 2)) &
+			   GPIO_ALTERNATE_MASK;
+
+	stm32_clock_disable(clock);
+}
+
+/* Apply GPIO (@bank/@pin) configuration described by @cfg */
+static void set_gpio_cfg(uint32_t bank, uint32_t pin, struct gpio_cfg *cfg)
+{
+	uintptr_t base = stm32_get_gpio_bank_base(bank);
+	unsigned int clock = stm32_get_gpio_bank_clock(bank);
+	uint32_t excep = cpu_spin_lock_xsave(&gpio_lock);
+
+	stm32_clock_enable(clock);
+
+	io_clrsetbits32(base + GPIO_MODER_OFFSET,
```

Comment:
This entire function is a bit hard to read. Some comments on what's done and some hints on the layout of various bitfields would help a lot.

Reply:
ok, i'll had few description.